### ISSUE TITLE
Error installing CloudBiolinux in a computing cluster

### ISSUE THREAD

**Issue by luiceur** [chapmanb/bcbio-nextgen#745]
Hi all,

We are trying to run bcbio from a computing cluster rather than from a cloud. Since it is a research cluster used by many other users we cannot have root access for the installation, therefore we have tried the isolated instalation method.
During the installation we get the following error related to CloudBiolinux:
```
Setting up virtual machine
INFO: <cloudbio.flavor.Flavor instance at 0x7fffd821b560>
INFO: This is a ngs_pipeline_minimal flavor
INFO: Reading default fabricrc.txt
DBG [config.py]: Using config file /home/U008/lcebaman/EdGen/tmpbcbio-install/cloudbiolinux/cloudbio/../config/fabricrc.txt
INFO: Distribution __auto__
INFO: Get local environment
Traceback (most recent call last):
  File "/home/U008/lcebaman/EdGen/bcbio/anaconda/bin/bcbio_nextgen.py", line 199, in <module>
    install.upgrade_bcbio(kwargs["args"])
  File "/home/U008/lcebaman/EdGen/bcbio/anaconda/lib/python2.7/site-packages/bcbio/install.py", line 75, in upgrade_bcbio
    upgrade_thirdparty_tools(args, REMOTES)
  File "/home/U008/lcebaman/EdGen/bcbio/anaconda/lib/python2.7/site-packages/bcbio/install.py", line 335, in upgrade_thirdparty_tools
    cbl_deploy.deploy(s)
  File "/home/U008/lcebaman/EdGen/tmpbcbio-install/cloudbiolinux/cloudbio/deploy/__init__.py", line 65, in deploy
    _setup_vm(options, vm_launcher, actions)
  File "/home/U008/lcebaman/EdGen/tmpbcbio-install/cloudbiolinux/cloudbio/deploy/__init__.py", line 101, in _setup_vm
    _setup_cloudbiolinux(options)
  File "/home/U008/lcebaman/EdGen/tmpbcbio-install/cloudbiolinux/cloudbio/deploy/__init__.py", line 193, in _setup_cloudbiolinux
    _configure_fabric_environment(env, flavor, fabricrc_loader=fabricrc_loader)
  File "/home/U008/lcebaman/EdGen/tmpbcbio-install/cloudbiolinux/cloudbio/utils.py", line 93, in _configure_fabric_environment
    _setup_distribution_environment(ignore_distcheck=ignore_distcheck)
  File "/home/U008/lcebaman/EdGen/tmpbcbio-install/cloudbiolinux/cloudbio/distribution.py", line 33, in _setup_distribution_environment
    env.distribution = _determine_distribution(env)
  File "/home/U008/lcebaman/EdGen/tmpbcbio-install/cloudbiolinux/cloudbio/distribution.py", line 297, in _determine_distribution
    raise Exception("Attempt to automatically determine Linux distribution of target machine failed, please manually specify distribution in fabricrc.txt")
Exception: Attempt to automatically determine Linux distribution of target machine failed, please manually specify distribution in fabricrc.txt
Traceback (most recent call last):
  File "bcbio_nextgen_install.py", line 255, in <module>
    main(parser.parse_args(), sys.argv[1:])
  File "bcbio_nextgen_install.py", line 41, in main
    subprocess.check_call([bcbio["bcbio_nextgen.py"], "upgrade"] + _clean_args(sys_argv, args, bcbio))
  File "/opt/anaconda/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/home/U008/lcebaman/EdGen/bcbio/anaconda/bin/bcbio_nextgen.py', 'upgrade',
'--tooldir=/home/U008/lcebaman/EdGen/bcbio', '--isolate', '--genomes', 'GRCh37', '--aligners', 'bwa', '--aligners', 'bowtie2', '--data']'
returned non-zero exit status 1``
```
which basically says that CouldBiolinux does not recognize the Linux distribution of our cluster. Our cluster is `SUSE Linux Enterprise Server 11 SP3  (x86_64) - Kernel \r (\l)` distribution and it does not seem to be one of the installation options for CouldBioLinux.
The installation process finishes there, however we can see that and `/bin` we have `bcbio_nextgen.py` and  `bcbio_setup_genome.py` scripts.
We have tried to run some samples and we obtained the following error:
```
File "/home/U008/lcebaman/EdGen/bcbio/anaconda/lib/python2.7/site-packages/bcbio/pipeline/genome.py", line 143, in _get_ref_from_galaxy_loc
    raise ValueError("Could not find reference genome file %s %s" % (genome_build, name))
ValueError: Could not find reference genome file GRCh37 samtools
```
Up to this point we have some worring questions and we were wondering if we could get some help here:
1- Do we need to install CouldBioLinux to be able to properly run bcbio?
2- If it is necessary since we want to run it locally and not in a cloud, how can we install CouldBioLinux in a cluster running Suse Linux Enterprise Server 11?
3- Is there any other way of bypassing the installation of CloudBioLinux?

We would really appreciate any help or ideas in this matter.

Best regards,
Luis

**Comment by chapmanb** [comment]
Luis;
Sorry about the issue. CloudBioLinux was not correctly detecting SUSE Linux, which is the cause of the error. If you remove any temporary cached install directories (`rm -rf tmpbcbio-install`) and re-run the same installer command it will pull down the latest version and hopefully work cleanly now. Sorry about the problem and hope it works now for you.

**Comment by luiceur** [comment]
Thank you very much for helping out here.
The installation managed to go through the Linux distribution so I think the fix worked. However we are now getting a `samtools` error:

```
  _install_pkg_latest(env, "samtools", [], brew_cmd, ipkgs)
  File "/home/U008/lcebaman/EdGen/tmpbcbio-install/cloudbiolinux/cloudbio/package/brew.py", line 205, in _install_pkg_latest
    raise ValueError("Failed to install brew formula: %s" % pkg)
ValueError: Failed to install brew formula: samtools
Traceback (most recent call last):
  File "bcbio_nextgen_install.py", line 255, in <module>
    main(parser.parse_args(), sys.argv[1:])
  File "bcbio_nextgen_install.py", line 41, in main
    subprocess.check_call([bcbio["bcbio_nextgen.py"], "upgrade"] + _clean_args(sys_argv, args, bcbio))
```
Have you come accross this problem before?

**Comment by chapmanb** [comment]
Luis;
Glad that helped get things further along and sorry about the continued build problems. It's tough to tell what is wrong from just this snippet, the useful error will be a bit upstream of this. If you can see the `samtools` brew command prior to this it would be useful to run that outside of the installer and add in a verbose flag (so `.../brew install` to `.../brew install -v`) and see what error that produces. That should indicate why the build fails.

**Comment by luiceur** [comment]
Brad;
This is what is happening here. The samtools `Makefile` uses `LIBCURSES=-lcurses` however in Suse and Fedora the curses library are called `-lncurses`. Now, I could easily install samtools on Suse modifying the `LIBCURSES` in the `Makefile`, however we don't know how to bypass this issue with `brew` since the package is downloaded and fresh installed from the samtiools main repository.
1- Could we skip the samtools installation from the main bcbio installation and then manually install samtools?
2- What other options do we have?

**Comment by chapmanb** [comment]
Luis;
Thanks for debugging this and figuring out the underlying issue. I proposed a patch upstream to homebrew-science which will enable use to avoid this by skipping the curses requirement. Once that gets pulled in, I can put in a long term fix in bcbio. In the short term, you can work around this by manually installing the new recipe:
```
wget https://raw.githubusercontent.com/chapmanb/homebrew-science/samtools-nocurses/samtools.rb
/path/to/bcbio/bin/brew install --without-curses -v samtools.rb
```
Then re-starting the install. It should know you have this in place now and not try to re-install it. Hope this helps and thanks again for the help debugging.

**Comment by luiceur** [comment]
Brad;
I managed to install samtools using `-lncurses` just modifying the formula and the package. However re-starting the installation is not helping. I believe the script keeps trying to install samtools and fails althought it is already installed.

**Comment by luiceur** [comment]
I have also tried eliminating `samtools` from  `/packages-homebrew.yaml` and re-starting the installation but it seems to keep downloading a new version of the script...

**Comment by chapmanb** [comment]
Luis;
Sorry about the continued issues. I'm confused as to why it would re-install samtools if you built it by hand with a custom recipe. Is this inside the brew installation for bcbio (`/home/U008/lcebaman/EdGen/bcbio/bin/brew`) or in some other way?

Either way I pushed a fix to cloudbiolinux that I hope will resolve this. If you remove the cached build directory (`rm -rf tmpbcbio-install`) and re-run it should now try to build samtools without the curses library and hopefully avoid the issue entirely. Please let us know if you run into any more problems. I hope this will get it running for you.

**Comment by luiceur** [comment]
Brad;
Thanks for helping us getting through the instalaltion process.  We have now passed `samtools` and found  a new problem this time with `freebayes`. This package installs fine on the system after modifying `fsom.h` library however it fails when it is installed with `bcbio_nextgen_install.py`. It also happens that it does not detect that the package has been already installed.
We do not plan to use `freebayes`, therefore a quick fix would be to be able to elimanate this package from the installation process:
1- Would that be possible?
2 -What other alternatives do we have?

**Comment by chapmanb** [comment]
Luis;
I pushed a fix that will hopefully resolve this issue if you remove any cached `tmpbcbio-install` directories and re-run. It should now skip the fsom build and hopefully avoid any issues.

On your more general questions:

- brew only looks at installations that occur via brew itself. So compiling outside of brew and putting in your path won't skip the build. It's definite useful for debugging but won't fix the underlying issue.
- We try to avoid custom changes like skipping tool builds. One of the goals is to have a community shared set of tools and if we have everyone doing custom workarounds we can't collect those in a central place to save future users time.

Sorry about the issues and hope this gets you further along in the build process.

**Comment by luiceur** [comment]
Brad;
Thank you very much for your help. Hopefully we can get the package installed soon.
I have deleted `tmpbcbio-install` directory and re-install it again but it keeps failing. Here is the error output of `brew -v install freebayes`:
```
[100%] Built target bamtools_cmd
make[1]: Leaving directory `/tmp/freebayes-nLZ66i/bamtools/build'
/home/U008/lcebaman/EdGen/bcbio/Cellar/cmake/3.1.1_1/bin/cmake -E cmake_progress_start /tmp/freebayes-nLZ66i/bamtools/build/CMakeFiles 0
==> make
g++ -w -O3 -pipe -fomit-frame-pointer -ffast-math fsom.c split.cpp main.cpp -o fsom -lm
In file included from fsom.c:21:
fsom.h:77: error: non-local function ‘void som_network_destroy(som_network_t*)’ uses anonymous type
fsom.h:75: error: ‘typedef struct<anonymous> som_network_t’ does not refer to the unqualified type, so it is not used for linkage
```
It seems to be pointing to `fsom.h`.

**Comment by chapmanb** [comment]
Luis;
Sorry, this is my mistake. I patched the Makefile but forgot we had a specific fsom build target in the recipe. I pushed a fix that removes this so we should now hopefully completely skip fsom. If you re-run fingers crossed this will fix it. Sorry about missing this.

**Comment by luiceur** [comment]
Brad;
Excellent! It worked! We have now an error with a package called `pindel`, but I think it is also due to the curses library again:
```
EdGen/bcbio/lib -Wl,-rpath,/home/U008/lcebaman/EdGen/bcbio/lib libbam.a -Lbcftools -lbcf  -lcurses  -lm -lz -lpthread
/usr/lib64/gcc/x86_64-suse-linux/4.3/../../../../x86_64-suse-linux/bin/ld: cannot find -lcurses
collect2: ld returned 1 exit status
make[1]: *** [samtools] Error 1
```
I think we are close now!!

**Comment by chapmanb** [comment]
Luis -- glad this is getting further and thanks again for all the reports and patience. I pushed a fix to the pindel build to avoid use of curses, so hopefully it'll proceed through that build if you re-run the install/update. Thanks again.

**Comment by luiceur** [comment]
Brad;
The installer seems to be hunging while fetching `esembl` during the `vep` installation:
```
export FTP_PASSIVE=1 && /home/U008/lcebaman/EdGen/bcbio/Cellar/vep/75_2014-06-12/INSTALL.pl -a a -d /home/U008/lcebaman/EdGen/bcbio/Cellar/vep/75_2014-06-12/lib -c /home/U008/lcebaman/EdGen/bcbio/Cellar/vep/75_2014-06-12/share/data
Using non-default installation directory /home/U008/lcebaman/EdGen/bcbio/Cellar/vep/75_2014-06-12/lib - you will probably need to add /home/U008/lcebaman/EdGen/bcbio/Cellar/vep/75_2014-06-12/lib to your PERL5LIB

Hello! This installer is configured to install v75 of the Ensembl API for use by the VEP.
It will not affect any existing installations of the Ensembl API that you may have.

It will also download and install cache files from Ensembl's FTP server.

Checking for installed versions of the Ensembl API...done

Setting up directories

Downloading required files
 - fetching ensembl
```
The installation does not go through and the job is killed by PBS after the walltime exceeded the limit

**Comment by chapmanb** [comment]
Luis;
Thanks for the additional reports. I've gone through and cleaned up the rest of the build issues on SUSE and hope this will now build cleanly for you. The issue with Ensembl hanging is due to this bug between Perls LWP downloader and the old version of openSSL shipped with SUSE 11:

https://rt.cpan.org/Public/Bug/Display.html?id=81575#txn-1153771

To fix, I swapped over the VEP installer to use curl by default. This does require having a working curl on your system. On `ultra` specifically, there is a broken curl installed along with anaconda in `/opt/anaconda/bin/curl`. It's worth either removing that curl globally (`/opt/anaconda/bin/conda uninstall curl`) or removing that directory from your path when running the install.

In addition to that I also added binaries for some tools like vt and WHAM that required more recent compilers than the 4.4 shipped with SUSE 11.

Hopefully it should now install cleanly if you fix the curl issue and re-run. Thanks again for all the patience and let me know if you run into anything else at all.
